app/controllers/api/users_controller.rb

```diff
@@ -1,7 +1,41 @@
+require "bcrypt"
+
 class Api::UsersController < Api::ApplicationController
+  wrap_parameters format: [:json]
+  before_action -> { doorkeeper_authorize! :account_data_read }, only: [:verify_pin]
   include Api::PublicApiExposingController
 
   def present
     @users = Presence.present_users
   end
+
+  def phone_number_token
+    secret = Rails.application.secrets.PHONE_TOKEN_SECRET
+    return head :internal_server_error if !secret
+
+    phone_app_id = Rails.application.secrets.PHONE_OAUTH_APP_ID
+    return head :internal_server_error if !phone_app_id
+
+    return head :bad_request if !params[:phone_number]
+    return head :forbidden if params[:secret] != secret
+
+    @user = User.joins(:phone_numbers).where(phone_numbers: {phone_number: params[:phone_number]}).limit(1).first
```

Comment:
Shouldn't we:
- limit to verified emails
- maybe we should order on created ASC, otherwise someone would be able to perform a denial of service

Reply:
Without a confirmed email the user cannot login to set a phone number and PIN anyway, so I don't see a point in checking for that.
Also, this requires an authentication token, and will be further restricted to localhost-only access from nginx, so I am not worried about DoS.

Reply:
It turned out that the phone numbers are unique